drivers/pktmem.c

```diff
@@ -97,86 +176,102 @@ static void net_freeDMAPbuf(struct pbuf *p)
 	SYS_ARCH_DECL_PROTECT(old_level);
 
 	SYS_ARCH_PROTECT(old_level);
-	net_freePktBuf(p->payload);
+	struct pktmem_pbuf_custom *ppc = (void *)p;
+	net_freePktBuf(p->payload, ppc->pktmem_start);
 	SYS_ARCH_UNPROTECT(old_level);
 
-	mem_free(p);
+	mem_free(ppc);
 }
 
 
-static ssize_t net_allocPktBuf(void **bufp)
+static ssize_t net_allocPktBuf(void **bufp, void **pktmem_start)
 {
 	SYS_ARCH_DECL_PROTECT(old_level);
-	buf_page_head_t *ph;
+	pktmem_info_t *info;
 	unsigned i;
 
 	SYS_ARCH_PROTECT(old_level);
-	if (pkt_bufs_free == 0) {
+	if (pkt_bufs_free == 0 || pktmem_lru.next == &pktmem_lru) {
 		SYS_ARCH_UNPROTECT(old_level);
 
-		ph = dmammap(PAGE_SIZE);
-		if (ph == NULL) {
+		info = dmammap(pktmem_opts.alloc_sz);
```

Comment:
Consider assigning the result to a new variable, to simplify the calculation at lines 225-226.